### PR TITLE
LGA-1766 - Pin all direct and indirect dependencies using pip-tools

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?

- Pin all direct and indirect dependencies using pip-tools
- Move direct dependency requirements files to requirements/source/*.in
- Add pre-commit hooks to compile requirements/source/*.in files to requirements/generated/*.txt
- Remove previous requirements/*.txt files
- Add check for compiling source files and checking they match generated files in the CI pipeline

## Checklist

- [X] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -88,6 +88,33 @@ jobs:
           command: |
             source lint-env/bin/activate
             black --check cla_backend
+  pip-compile:
+    docker:
+      - image: circleci/python:2.7
```
Comment (jogilder) [inline .circleci/config.yml:93]: I think this image might be deprecated in favour of cimg images. [https://circleci.com/developer/images/image/cimg/python](url)
[https://discuss.circleci.com/t/legacy-convenience-image-deprecation/41034](url)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,3 +9,26 @@ repos:
     hooks:
     - id: flake8
       args: ['--config=setup.cfg',  '--exclude=./cla_backend/settings/*,./docs/*']
+-   repo: https://github.com/jazzband/pip-tools
```
Comment (jogilder) [inline .pre-commit-config.yaml:12]: what happens if you edit requirements-base.in? You mentioned it in the training.

Comment (said-moj) [reply .pre-commit-config.yaml:12]: @jogilder I did further work after the video to make sure it recompiles after base.in is changed, see this commit https://github.com/ministryofjustice/cla_backend/pull/838/commits/7981e31716420644b6188891251aee3dfdc68d7f

File: .circleci/config.yml (hunk 1 above)
Comment (said-moj) [reply .circleci/config.yml]: Good spot,I have now changed this to use the cimg image

**Review by jogilder — APPROVED** [review]
LGTM

**Pull request merged**